### PR TITLE
Refactor sandbox module cache

### PR DESCRIPTION
**Pull request by javivelasco**
To run middleware we are using a **sandbox** that emulates the web runtime and keeps a module cache. This cache is shared for all of the modules that we run using the sandbox while there are some module-level APIs that must be scoped depending on the module we are running.

One example of this is `fetch` where we want to always inject a special header that indicate the module that is performing the fetch and use it to avoid getting into infinite loops for middleware. For those cases the cached implementation will be the first one that instantiates the module and therefore we can actually get into infinite loops. This is the reason why #31800 is failing.

With this PR we refactor the sandbox so that the module cache is scoped per module name. This means that one execution of a middleware will preserve its cache only for that module so that each execution will still have its own `fetch` implementation, fixing this issue. Also, with this refactor the code is more clear and we also provide an option to avoid using the cache.

## Bug

- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Errors have helpful link attached, see `contributing.md`

## Feature

- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [ ] Make sure the linting passes by running `yarn lint`

### CODE REVIEW

File: packages/next/server/web/sandbox/context.ts [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+    // Keyed collections
+    Map,
+    Set,
+    WeakMap,
+    WeakSet,
+
+    // Structured data
+    ArrayBuffer,
+    SharedArrayBuffer,
```
Comment (sokra) [inline packages/next/server/web/sandbox/context.ts:224]: Why are these listed here? shouldn't the context have it's own variants of them?

File: packages/next/server/web/sandbox/context.ts [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+  onWarning: (warn: Error) => void
+  useCache: boolean
+}) {
+  const cache = options.useCache
+    ? caches.get(options.module) || createModuleContext(options)
+    : createModuleContext(options)
+
+  if (options.useCache) {
+    caches.set(options.module, cache)
+  }
```
Comment (sokra) [inline packages/next/server/web/sandbox/context.ts:62]:
```suggestion
  const cache = options.useCache
    ? caches.get(options.module)
    : createModuleContext(options)

  if (cache === undefined) {
    caches.set(options.module, cache = createModuleContext(options))
  }
```

This calls `caches.set` only when needed

File: packages/next/server/web/sandbox/context.ts [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+  const moduleCache = {
+    context: context,
+    paths: new Map<string, string>(),
+    require: requireCache,
+    warnedEvals: new Set<string>(),
+  }
+
+  context.__next_eval__ = function __next_eval__(fn: Function) {
+    const key = fn.toString()
+    if (!moduleCache?.warnedEvals.has(key)) {
```
Comment (sokra) [inline packages/next/server/web/sandbox/context.ts:130]:
```suggestion
    if (!moduleCache.warnedEvals.has(key)) {
```

File: packages/next/server/web/sandbox/context.ts [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+      const warning = new Error(
+        `Dynamic Code Evaluation (e. g. 'eval', 'new Function') not allowed in Middleware`
+      )
+      warning.name = 'DynamicCodeEvaluationWarning'
+      Error.captureStackTrace(warning, __next_eval__)
+      moduleCache?.warnedEvals.add(key)
```
Comment (sokra) [inline packages/next/server/web/sandbox/context.ts:136]:
```suggestion
      moduleCache.warnedEvals.add(key)
```

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| buildDuration | 21.7s | 21.2s | -536ms |
| buildDurationCached | 4s | 4s | -2ms |
| nodeModulesSize | 345 MB | 345 MB | ⚠️ +2.72 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.726 | 3.696 | -0.03 |
| / avg req/sec | 670.91 | 676.38 | +5.47 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.865 | 1.845 | -0.02 |
| /error-in-render avg req/sec | 1340.73 | 1355.12 | +14.39 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28.4 kB | 28.4 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.45 kB | 4.45 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.3 kB | 13.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| buildDuration | 24.6s | 23.3s | -1.3s |
| buildDurationCached | 3.9s | 4s | ⚠️ +111ms |
| nodeModulesSize | 345 MB | 345 MB | ⚠️ +2.72 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.817 | 3.49 | -0.33 |
| / avg req/sec | 655.04 | 716.4 | +61.36 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.874 | 1.894 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 1334.07 | 1319.65 | ⚠️ -14.42 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.37 kB | 2.37 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.47 kB | 4.47 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.2 kB | 13.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 996246538137ed66295fef08277b0f607fb1ea0e

**Review by sokra — DISMISSED** [review]
a few nitpicks, otherwise fine

File: packages/next/server/web/sandbox/context.ts (hunk 1 above)
Comment (javivelasco) [reply packages/next/server/web/sandbox/context.ts:224]: It was already there before but I think it's simply to be explicit. If at some point we want to tweak one of those implementations we can simply substitute it there.

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| buildDuration | 17.4s | 17.5s | ⚠️ +59ms |
| buildDurationCached | 3.1s | 3.1s | ⚠️ +27ms |
| nodeModulesSize | 345 MB | 345 MB | ⚠️ +2.52 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall increase ✓</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.812 | 2.809 | 0 |
| / avg req/sec | 889.17 | 890.09 | +0.92 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.363 | 1.336 | -0.03 |
| /error-in-render avg req/sec | 1834 | 1870.78 | +36.78 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 28.4 kB | 28.4 kB | ✓ |
| webpack-HASH.js gzip | 1.45 kB | 1.45 kB | ✓ |
| Overall change | 72.2 kB | 72.2 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.37 kB | 1.37 kB | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 327 B | 327 B | ✓ |
| dynamic-HASH.js gzip | 2.38 kB | 2.38 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 635 B | 635 B | ✓ |
| image-HASH.js gzip | 4.45 kB | 4.45 kB | ✓ |
| index-HASH.js gzip | 263 B | 263 B | ✓ |
| link-HASH.js gzip | 1.87 kB | 1.87 kB | ✓ |
| routerDirect..HASH.js gzip | 321 B | 321 B | ✓ |
| script-HASH.js gzip | 383 B | 383 B | ✓ |
| withRouter-HASH.js gzip | 318 B | 318 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.3 kB | 13.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 460 B | 460 B | ✓ |
| Overall change | 460 B | 460 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| index.html gzip | 532 B | 532 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Default Build with SWC</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| buildDuration | 18.9s | 19.3s | ⚠️ +335ms |
| buildDurationCached | 3.2s | 3.2s | ⚠️ +65ms |
| nodeModulesSize | 345 MB | 345 MB | ⚠️ +2.52 kB |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.737 | 2.958 | ⚠️ +0.22 |
| / avg req/sec | 913.56 | 845.05 | ⚠️ -68.51 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.242 | 1.351 | ⚠️ +0.11 |
| /error-in-render avg req/sec | 2012.08 | 1850.93 | ⚠️ -161.15 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| 450.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.3 kB | 42.3 kB | ✓ |
| main-HASH.js gzip | 28.6 kB | 28.6 kB | ✓ |
| webpack-HASH.js gzip | 1.44 kB | 1.44 kB | ✓ |
| Overall change | 72.5 kB | 72.5 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31 kB | 31 kB | ✓ |
| Overall change | 31 kB | 31 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _app-HASH.js gzip | 1.35 kB | 1.35 kB | ✓ |
| _error-HASH.js gzip | 180 B | 180 B | ✓ |
| amp-HASH.js gzip | 305 B | 305 B | ✓ |
| css-HASH.js gzip | 321 B | 321 B | ✓ |
| dynamic-HASH.js gzip | 2.37 kB | 2.37 kB | ✓ |
| head-HASH.js gzip | 342 B | 342 B | ✓ |
| hooks-HASH.js gzip | 622 B | 622 B | ✓ |
| image-HASH.js gzip | 4.47 kB | 4.47 kB | ✓ |
| index-HASH.js gzip | 256 B | 256 B | ✓ |
| link-HASH.js gzip | 1.91 kB | 1.91 kB | ✓ |
| routerDirect..HASH.js gzip | 314 B | 314 B | ✓ |
| script-HASH.js gzip | 375 B | 375 B | ✓ |
| withRouter-HASH.js gzip | 309 B | 309 B | ✓ |
| 85e02e95b279..7e3.css gzip | 107 B | 107 B | ✓ |
| Overall change | 13.2 kB | 13.2 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| _buildManifest.js gzip | 458 B | 458 B | ✓ |
| Overall change | 458 B | 458 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | javivelasco/next.js refactor/sandbox-cache | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 545 B | 545 B | ✓ |
| withRouter.html gzip | 526 B | 526 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
Commit: 3bdac83772a03132a626ecfda0312048e9f46f29

**Review by timneutkens — APPROVED** [review]

**Pull request merged**